File: leetcode/704.binary-search/go/binarysearch.go
package binarysearch

func search(nums []int, target int) int {
	return searchBruteForce(nums, target)
}

// LogN, LogN
func searchOptimized(nums []int, target int) int {
	return searchHelper(nums, target, 0, len(nums))
}

func searchHelper(nums []int, target int, start int, end int) int {
	if end <= start {
		return -1
	}

	mid := start + (end-start)/2
	midValue := nums[mid]

	if target == midValue {
		return mid
	}

	if target < midValue {
		return searchHelper(nums, target, start, mid)
	}

	return searchHelper(nums, target, mid+1, end)
}

// N, 1
func searchBruteForce(nums []int, target int) int {
	for i, v := range nums {
		if v == target {
			return i
		}
	}

	return -1
}

File: leetcode/704.binary-search/ts/binarysearch.ts
const search = (nums: number[], target: number): number => {
    return searchOptimized(nums, target)
};

// LogN, logN
const searchOptimized = (nums: number[], target: number): number => {
    return searchHelper(nums, target, 0, nums.length)
}

const searchHelper = (nums: number[], target: number, start: number, end: number): number => {
    if (start >= end) return -1

    const mid = Math.floor(start + (end - start) / 2)
    const midValue = nums[mid]

    if (midValue === target) return mid

    if (target < midValue) return searchHelper(nums, target, start, mid);

    return searchHelper(nums, target, mid + 1, end)
}

// N, 1
const searchBruteForce = (nums: number[], target: number): number => {
    return nums.findIndex(n => n === target)
}

export default search;